File: arr.cpp
#include "arr.h"
#include <iostream>
#include <fstream>
#include <cassert>

using namespace std;

void MyArr::ensureCapacity(int newSize) {
    if (newSize <= capacity) return;
    
    int newCapacity = (capacity == 0) ? 4 : capacity * 2;
    while (newCapacity < newSize) newCapacity *= 2;
    
    string* newData = new string[newCapacity];
    for (int i = 0; i < size; ++i) {
        newData[i] = data[i];
    }
    
    delete[] data;
    data = newData;
    capacity = newCapacity;
}

MyArr::MyArr() : data(nullptr), size(0), capacity(0) {}

MyArr::~MyArr() {
    delete[] data;
}

void MyArr::addHead(const string& val) {
    ensureCapacity(size + 1);
    for (int i = size; i > 0; --i) {
        data[i] = data[i - 1];
    }
    data[0] = val;
    ++size;
}

void MyArr::delHead() {
    if (size == 0) return;
    for (int i = 0; i < size - 1; ++i) {
        data[i] = data[i + 1];
    }
    --size;
}

void MyArr::delEnd() {
    if (size == 0) return;
    --size;
}

void MyArr::addEnd(const string& val) {
    ensureCapacity(size + 1);
    data[size] = val;
    ++size;
}

string MyArr::getAt(int idx) const {
    if (idx < 0 || idx >= size) return "[INVALID_INDEX]";
    return data[idx];
}

void MyArr::delAt(int idx) {
    if (idx < 0 || idx >= size) return;
    for (int i = idx; i < size - 1; ++i) {
        data[i] = data[i + 1];
    }
    --size;
}

void MyArr::readArray() const {
    if (size == 0) {
        cout << "Массив пуст." << endl;
        return;
    }
    cout << "Массив [" << size << "]: ";
    for (int i = 0; i < size; ++i) {
        cout << data[i];
        if (i < size - 1) cout << " -> ";
    }
    cout << endl;
}

void MyArr::addAt(int idx, const string& val) {
    if (idx < 0 || idx > size) return;
    ensureCapacity(size + 1);
    for (int i = size; i > idx; --i) {
        data[i] = data[i - 1];
    }
    data[idx] = val;
    ++size;
}

void MyArr::repArr(int idx, const string& val) {
    if (idx < 0 || idx >= size) return;
    data[idx] = val;
}

auto MyArr::lenArr() const -> int {
    return size;
}

File: ds/arr.go
package ds

import "fmt"

type MyArr struct {
	data []string
}

func NewMyArr() *MyArr {
	return &MyArr{
		data: make([]string, 0),
	}
}

func (a *MyArr) AddEnd(val string) {
	a.data = append(a.data, val)
}

func (a *MyArr) AddHead(val string) {
	a.data = append([]string{val}, a.data...)
}

func (a *MyArr) AddAt(idx int, val string) {
	if idx < 0 || idx > len(a.data) {
		return
	}
	if idx == len(a.data) {
		a.AddEnd(val)
		return
	}
	a.data = append(a.data[:idx+1], a.data[idx:]...)
	a.data[idx] = val
}

func (a *MyArr) DelHead() {
	if len(a.data) == 0 {
		return
	}
	a.data = a.data[1:]
}

func (a *MyArr) DelEnd() {
	if len(a.data) == 0 {
		return
	}
	a.data = a.data[:len(a.data)-1]
}

func (a *MyArr) DelAt(idx int) {
	if idx < 0 || idx >= len(a.data) {
		return
	}
	a.data = append(a.data[:idx], a.data[idx+1:]...)
}

func (a *MyArr) RepArr(idx int, val string) {
	if idx < 0 || idx >= len(a.data) {
		return
	}
	a.data[idx] = val
}

func (a *MyArr) GetAt(idx int) string {
	if idx >= 0 && idx < len(a.data) {
		return a.data[idx]
	}
	return "[INVALID_INDEX]"
}

func (a *MyArr) LenArr() int {
	return len(a.data)
}

func (a *MyArr) ReadArray() {
	if len(a.data) == 0 {
		fmt.Println("Массив пуст.")
		return
	}
	fmt.Printf("Содержимое массива (%d/%d): ", len(a.data), cap(a.data))
	for _, val := range a.data {
		fmt.Printf("\"%s\" ", val)
	}
	fmt.Println()
}